libs/nio/src/main/java/org/elasticsearch/nio/SocketChannelContext.java

```diff
@@ -234,49 +234,107 @@ protected boolean closeNow() {
+    // copied again on the next call.
+
     protected int readFromChannel(ByteBuffer buffer) throws IOException {
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+        ioBuffer.limit(Math.min(buffer.remaining(), ioBuffer.limit()));
+        int bytesRead;
         try {
-            int bytesRead = rawChannel.read(buffer);
-            if (bytesRead < 0) {
-                closeNow = true;
-                bytesRead = 0;
-            }
-            return bytesRead;
+            bytesRead = rawChannel.read(ioBuffer);
         } catch (IOException e) {
             closeNow = true;
             throw e;
         }
+        if (bytesRead < 0) {
+            closeNow = true;
+            return 0;
+        } else {
+            ioBuffer.flip();
+            buffer.put(ioBuffer);
+            return bytesRead;
+        }
     }
 
-    protected int readFromChannel(ByteBuffer[] buffers) throws IOException {
+    protected int readFromChannel(InboundChannelBuffer channelBuffer) throws IOException {
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+        int bytesRead;
         try {
-            int bytesRead = (int) rawChannel.read(buffers);
-            if (bytesRead < 0) {
-                closeNow = true;
-                bytesRead = 0;
-            }
-            return bytesRead;
+            bytesRead = rawChannel.read(ioBuffer);
         } catch (IOException e) {
             closeNow = true;
             throw e;
         }
+        if (bytesRead < 0) {
+            closeNow = true;
+            return 0;
+        } else {
+            ioBuffer.flip();
+            channelBuffer.ensureCapacity(channelBuffer.getIndex() + ioBuffer.remaining());
+            ByteBuffer[] buffers = channelBuffer.sliceBuffersFrom(channelBuffer.getIndex());
+            int j = 0;
+            while (j < buffers.length && ioBuffer.remaining() > 0) {
+                ByteBuffer buffer = buffers[j++];
+                copyBytes(ioBuffer, buffer);
+            }
+            channelBuffer.incrementIndex(bytesRead);
+            return bytesRead;
+        }
     }
 
     protected int flushToChannel(ByteBuffer buffer) throws IOException {
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+        copyBytes(buffer, ioBuffer);
+        ioBuffer.flip();
         try {
-            return rawChannel.write(buffer);
+            return rawChannel.write(ioBuffer);
         } catch (IOException e) {
             closeNow = true;
             throw e;
         }
     }
 
     protected int flushToChannel(ByteBuffer[] buffers) throws IOException {
-        try {
-            return (int) rawChannel.write(buffers);
-        } catch (IOException e) {
-            closeNow = true;
-            throw e;
+        ByteBuffer ioBuffer = getSelector().getIoBuffer();
+
+        boolean continueFlush = buffers.length != 0;
+        int bytesFlushed = 0;
+        int lastBufferCopied = 0;
+        while (continueFlush) {
```

Comment:
This loop looks like we're turning a non-blocking write into a blocking one. If a channel is not writable for some period of time (maybe become we're passing a lot of data in `buffers`) won't this approach block the IO thread?

Reply:
Good catch. That should be checking that greater than 0 was flushed. I'll fix it.

Reply:
Good catch. That should be checking that greater than 0 was flushed. I'll fix it.

Reply:
Actually no. If we have any remaining data in the io buffer after flush we stop looping.

Reply:
Right my bad sorry ... misread the last line in the loop.